src/estimators/meta_estimators.py

```diff
@@ -0,0 +1,171 @@
+  is an exact set, this estimator is equivalent to the Bloom Filter base
+  estimator.
+  """
+
+  def __init__(self,
+               meta_sketch_factory,
+               meta_sketch_estimator,
+               adbf_estimator):
+    """Initializes a MetaEstimator using the given FirstMomentEstimator method.
+
+    Args:
+      meta_sketch_factory: a sketch factory method used to estimate Bloom Filter
+        buckets.
+      meta_sketch_estimator: an estimator to estimate the union cardinality
+        given a list of meta sketches.
+      adbf_estimator: an estimator to estimate the cardinality of a list of
+        AnyDistributionBloomFilter sketches.
+    """
+    super().__init__()
+    self.adbf_estimator = adbf_estimator
+    self.meta_sketch_factory = meta_sketch_factory
+    self.meta_sketch_estimator = meta_sketch_estimator
+
+  @classmethod
+  def _check_compatibility(cls, sketch_list):
+    """Determines if all sketches are compatible."""
+    for sketch in sketch_list:
+      assert isinstance(sketch, bf.AnyDistributionBloomFilter), (
+          'MetaEstimator must be used with an AnyDistributionBloomFilter.')
+
+    first_sketch = sketch_list[0]
+
+    for cur_sketch in sketch_list[1:]:
+      first_sketch.assert_compatible(cur_sketch)
+
+  def __call__(self, sketch_list):
+    """Transform AnyDistributionBloomFilters into meta sketch and estimate union."""
+    MetaEstimatorBase._check_compatibility(sketch_list)
+    if not sketch_list:
+      return 0
+
+    meta_sketches = self._transform_adbf_into_meta_sketches(sketch_list)
+    num_ones = self._estimate_num_ones(meta_sketches)
+    adbf_sketch = MetaEstimatorBase._construct_fake_adbf(
+        num_ones, sketch_list[0])
+    return self.adbf_estimator([adbf_sketch])
+
+  def _transform_adbf_into_meta_sketches(self, sketch_list):
+    """Transform AnyDistributionBloomFilters into meta sketch.
+
+    Args:
+      sketch_list: a list of AnyDistributionBloomFilter sketches.
+
+    Returns:
+      A list of meta sketches.
+    """
+    if not sketch_list:
+      return []
+
+    random_seed = sketch_list[0].random_seed
+
+    meta_sketches = []
+    for sketch in sketch_list:
+      meta_sketch = self.meta_sketch_factory(random_seed)
+
+      # Get nonzero indices in sketch and add those indices to the meta_sketch.
+      nonzero_indices = np.flatnonzero(sketch.sketch)
+      meta_sketch.add_ids(nonzero_indices)
+
+      meta_sketches.append(meta_sketch)
+
+    return meta_sketches
+
+  def _estimate_num_ones(self, meta_sketches):
+    """Estimate the number of one registers from the meta sketches.
+
+    Args:
+      meta_sketches: a list of meta sketches.
+
+    Returns:
+      The estimated number of one registers.
+    """
+    # Estimate the number of 1's from the meta sketches.
+    return int(self.meta_sketch_estimator(meta_sketches)[0])
+
+  @classmethod
+  def _construct_fake_adbf(cls, num_ones, template_adbf):
```

Comment:
This is clever!

Reply:
This is Prestons idea. Really brilliant!